File: src/main/java/com/alphawallet/attestation/eip712/Timestamp.java
package com.alphawallet.attestation.eip712;


import com.alphawallet.attestation.core.ExceptionUtil;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.Clock;
import java.util.Date;
import java.util.Locale;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class Timestamp {
  private static final Logger logger = LogManager.getLogger(Timestamp.class);

  public static final int ALLOWED_ROUNDING = 1000; // 1 sec, since we are always rounding to the nearest second in the string representation
  // Timestamp with millisecond accuracy and timezone info
  public static final SimpleDateFormat TIMESTAMP_FORMAT = new SimpleDateFormat("EEE MMM d yyyy HH:mm:ss 'GMT'Z", Locale.US);

  // See RFC 5282, https://tools.ietf.org/html/rfc5280#section-4.1.2.5, based on the GeneralizedTime value of 99991231235959Z
  public static final long UNLIMITED = 253402297199000L;
  public static final long DEFAULT_TOKEN_TIME_LIMIT = 1000 * 60 * 60 * 24 * 365; // 1 year
  public static final long DEFAULT_TIME_LIMIT_MS = 1000*60*20; // 20 minutes

  private final long time;
  private long validity = 0;

  public Timestamp() {
    long tempTime = getCurrentTime();
    // Round down to nearest second
    this.time = tempTime - (tempTime % 1000);
  }

  public Timestamp(long timeSinceEpochInMs) {
    // Round down to nearest second
    this.time = timeSinceEpochInMs - (timeSinceEpochInMs % 1000);
  }

  public Timestamp(String timeAsString) {
    this.time = stringTimestampToLong(timeAsString);
  }

  public long getValidity() {
    return validity;
  }

  public void setValidity(long validity) {
    this.validity = validity;
  }

  public long getTime() {
    return time;
  }

  public String getTimeAsString() {
    return Timestamp.TIMESTAMP_FORMAT.format(new Date(time));
  }

  public boolean validateTimestamp() {
    long currentTime = getCurrentTime();
    if (time > currentTime + ALLOWED_ROUNDING) {
      logger.error("Timestamp is from the future");
      return false;
    }
    // Slack only goes into the future
    if (time < currentTime - ALLOWED_ROUNDING - validity) {
      logger.error("Timestamp is expired");
      return false;
    }
    return true;
  }

  public boolean validateAgainstExpiration(long expirationTimeInMs) {
    long currentTime = getCurrentTime();
    // If timestamp is in the future
    if (time > currentTime + ALLOWED_ROUNDING) {
      logger.error("Timestamp is from the future");
      return false;
    }
    // If token has expired
    if (expirationTimeInMs < currentTime - ALLOWED_ROUNDING) {
      logger.error("Expiration time has passed");
      return false;
    }
    // If the token is valid for too long
    if (expirationTimeInMs - time > validity + ALLOWED_ROUNDING) {
      logger.error("Lifetime is larger than allowed");
      return false;
    }
    return true;
  }

  public static long stringTimestampToLong(String timestamp) {
    try {
      return TIMESTAMP_FORMAT.parse(timestamp).getTime();
    } catch (ParseException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode timestamp", e);
    }
  }

  protected long getCurrentTime() {
    return Clock.systemUTC().millis();
  }
}

File: src/main/javascript/crypto/src/libs/Timestamp.ts

export class Timestamp {
    // Timestamp with millisecond accuracy and timezone info
    // Date.parse('Thu, 01 Jan 1970 00:00:00 GMT-0400');
    // Tue Mar 30 2021 21:14:22 GMT+0300
    public static TIMESTAMP_FORMAT:string = "EEE MMM d yyyy HH:mm:ss 'GMT'Z";
    public ALLOWED_ROUNDING: number = 1000; // 1 sec, since we are always rounding to the nearest second in the string representation
    static ALLOWED_ROUNDING: number = 1000; // 1 sec, since we are always rounding to the nearest second in the string representation

    static UNLIMITED = 253402297199000;
    static DEFAULT_TOKEN_TIME_LIMIT = 1000 * 60 * 60 * 24 * 365; // 1 year
    static DEFAULT_TIME_LIMIT_MS = 1000*60*20; // 20 minutes

    private time: number;
    private validity: number = 0;

    public constructor(timeSinceEpochInMs:number|string = null) {
        if (!timeSinceEpochInMs) this.time = Date.now();

        if (typeof timeSinceEpochInMs === 'number' ){
            this.time = timeSinceEpochInMs;
        }

        if (typeof timeSinceEpochInMs === 'string' ){
            this.time = Timestamp.stringTimestampToLong(timeSinceEpochInMs);
        }

        this.time = this.time - this.time % 1000;
    }


    public fromString(timeAsString: string) {
        this.time = Timestamp.stringTimestampToLong(timeAsString);
    }

    public getValidity():number {
        return this.validity;
    }

    public setValidity(validity: number) {
        this.validity = validity;
    }

    public getTime(): number {
        return this.time;
    }

    public getTimeAsString(): string {
        let preTime = new Date(this.time).toString();
        return preTime.substr(0, preTime.indexOf('(') - 1);
    }

    public validateTimestamp(): boolean {
        let currentTime = this.getCurrentTime();
        if (this.time > currentTime + this.ALLOWED_ROUNDING) {
            return false;
        }
        // Slack only goes into the future
        if (this.time < currentTime - this.ALLOWED_ROUNDING - this.validity) {
            return false;
        }
        return true;
    }

    public validateAgainstExpiration(expirationTimeInMs: number): boolean {
        let currentTime = this.getCurrentTime();
        // If timestamp is in the future
        if (this.time > (currentTime + this.ALLOWED_ROUNDING)) {
            return false;
        }
        // If token has expired
        if (expirationTimeInMs < (currentTime - this.ALLOWED_ROUNDING)) {
            return false;
        }
        // If the token is valid for too long
        if ((expirationTimeInMs - this.time) > (this.validity + this.ALLOWED_ROUNDING)) {
            console.log(expirationTimeInMs + "\n" + this.time + "\n" + this.validity + "\n" + this.ALLOWED_ROUNDING + "\n" + (expirationTimeInMs - this.time)  + "\n" + (this.validity + this.ALLOWED_ROUNDING) + "\n" );
            return false;
        }
        return true;
    }

    public static stringTimestampToLong(timestamp: string):number {
        return Date.parse(timestamp);
    }

    protected getCurrentTime(): number {
        return Date.now();
    }
}
